complete support for RFC 31 Job Constraints

grondo opened the pull request:
This PR is a WIP. The current solution almost works, but there is a compilation error discussed on slack in `resource/utilities/resource-query.cpp` due to the addition of a `Flux::resource_model::resource_t` base class.

```
  CXX      resource_query-resource-query.o
resource-query.cpp: In function ‘void write_to_graphviz(Flux::resource_model::f_resource_graph_t&, Flux::resource_model::subsystem_t, std::fstream&)’:
resource-query.cpp:319:33: error: conversion from ‘vec_adj_list_vertex_property_map<[...],[...],[...],[...],std::__cxx11::basic_string<char> Flux::resource_model::resource_t::*>’ to non-scalar type ‘vec_adj_list_vertex_property_map<[...],[...],[...],[...],std::__cxx11::basic_string<char> Flux::resource_model::resource_pool_t::*>’ requested
  319 |     f_res_name_map_t vmap = get (&resource_pool_t::name, fg);
      |                             ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm posting the PR early to not only get help fixing or working around this error, but to get any feedback on the implementation so far.

Summary:

This PR adds more complete support for RFC 31 Job Constraints specified in jobspec. Currently, Fluxion only supports constraints of the exact form `{ "properties": [ "xx", "yy" ] }`, however to support constraints on hosts, ranks, etc we will need to support the more full RFC 31 (currently optional) set of "operators" including the boolean operators `and` `or` and `not`.

This PR adds a new `Jobspec::Constraint` base class and constraint "parser" which parses the constraints section of jobspec and, if valid, returns a `Constraint *` which can then be used to determine if the constraint matches a resource vertex.

Each operator is implemented as a derived class of `Constraint`, e.g. a `PropertyConstraint` contains a vector of properties and its `match()` method handles checking that all specified properties are set in the resource, and the `ConditionalConstraint` contains a vector of other `Constraint` objects and returns the appropriate result if all/none/one match (`and`/`not`/`or`).

Later, we can add a `HostlistConstraint` and `RankConstraint` to support constraining jobs to hosts/ranks and/or excluding hosts/ranks.

The `constraint_parser()` function returns a `shared_ptr` because the Constraint has to be shared between the jobspec class and the `jobmeta` class used in the traverser.

Fixes #942
Fixes #965

grondo commented:
Ok, thanks to help from @milroy and @trws we found and fixed the issue with `resource-query.cpp` and all tests are passing.

I'm removing WIP so that this can get a more thorough review, though I still have to expand some testing.

I also have a plan to tweak the design slightly so that a Constraint can be matched to any resource type (since previous to this properties on non-node resources were ostensibly supported by the code, though it wasn't tested nor used anywhere)

grondo commented:
Ok, this should be ready for a full review.

> I also have a plan to tweak the design slightly so that a Constraint can be matched to any resource type (since previous to this properties on non-node resources were ostensibly supported by the code, though it wasn't tested nor used anywhere)

I ran into a bit of a snag with my idea, but since there isn't a current use case for this I think we can kick the can down the road a bit.

grondo commented:
demo using my flux-core branch that supports RFC 35 constraint query syntax:
```
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux resource list
     STATE PROPERTIES NNODES   NCORES    NGPUS NODELIST
      free batch           2        8        0 fluke[6-7]
      free xx,batch        2        8        0 fluke[8-9]
 allocated                 0        0        0 
      down                 0        0        0 
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux mini run --requires=rank:0 hostname
fluke6
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux mini run --requires=rank:2 hostname
fluke8
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux mini run --requires=host:fluke[7,8] hostname
fluke8
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux mini run --requires=-host:fluke[7,8] hostname
fluke9
f(s=4,d=1,builddir) grondo@fluke6:~/git/flux-core.git$ flux mini run --requires='host:fluke[7,8] and xx' hostname
fluke8
```

milroy commented on resource/libjobspec/rank_constraint.cpp:
```diff
@@ -0,0 +1,59 @@
+        if (add_idset_string (ranks, ids.c_str()) < 0) {
+            idset_destroy (ranks);
+            std::string msg = "Invalid idset `" + ids + "'";
+            throw parse_error (val, msg.c_str());
```
Spacing nit:
```suggestion
            throw parse_error (val, msg.c_str ());
```

milroy commented on resource/libjobspec/rank_constraint.cpp:
```diff
@@ -0,0 +1,59 @@
+        throw parse_error (values, "Out of memory");
+    for (auto&& val: values) {
+        std::string ids = val.as<std::string>();
+        if (add_idset_string (ranks, ids.c_str()) < 0) {
```
Spacing nit:
```suggestion
        if (add_idset_string (ranks, ids.c_str ()) < 0) {
```

milroy commented on resource/libjobspec/rank_constraint.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (!(ranks = idset_create (0, IDSET_FLAG_AUTOGROW)))
+        throw parse_error (values, "Out of memory");
+    for (auto&& val: values) {
+        std::string ids = val.as<std::string>();
```
Spacing nit:
```suggestion
        std::string ids = val.as<std::string> ();
```

grondo commented on resource/Makefile.am:
```diff
@@ -56,6 +56,7 @@ libresource_la_SOURCES = \
     schema/data_std.hpp \
```
Good catch, thanks! It feels strange to say we're deriving a class from a struct, so I'll reword to something like:

>  Define a new resource_t struct to act as a base resource class.

grondo commented on resource/libjobspec/constraint.cpp:
```diff
@@ -0,0 +1,211 @@
+
+class ConditionalConstraint : public Constraint {
+public:
+    ConditionalConstraint (std::string&, const YAML::Node&);
```
Good point, will fix.

grondo commented on resource/libjobspec/constraint.cpp:
```diff
@@ -0,0 +1,211 @@
+        return std::shared_ptr<Constraint>(new Constraint ());
+
+    it = constraint.begin();
+    std::string operation = it->first.as<std::string>();
```
Good catch! This is also codified in [RFC 7](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_7.html#flux-coding-style-guide).

milroy replied:
Interesting; I always thought it was a convention unique to Flux. I didn't know this: 

> One space SHALL separate function names and the opening parenthesis (enhances readability).
was from Kernighan & Ritchie coding style.

grondo commented:
Thanks @milroy! I've cleaned up the whitespace errors and the noted commit message typo and ambiguity and forced a push.

milroy commented on resource/libjobspec/test/constraint.cpp:
```diff
@@ -0,0 +1,250 @@
+
+int main (int ac, char *argv[])
+{
+    plan (NO_PLAN);
```
Never mind. The answer is because `ok` is called in loops.

milroy commented on resource/libjobspec/constraint.cpp:
```diff
@@ -0,0 +1,217 @@
+
+private:
+    std::string op;
+    std::vector<std::shared_ptr<Constraint>> values;
```
I typed this question in the firs review, but somehow it disappeared: so I understand the class design better, how will the ownership of `values` be shared?

grondo replied:
It shouldn't. A unique_ptr should work here, but I thought I had tried that and got compilation errors.
However, let me retry that, since the code was quite different when I gave up on unique_ptr.

grondo replied:
@milroy: See my response to @trws below. I think you are correct that this should be a vector of `unique_ptr`, but I'm not sure how to accomplish that given that the general Constraint factory function needs to return `shared_ptr`.

milroy commented on resource/libjobspec/hostlist_constraint.hpp:
```diff
@@ -0,0 +1,39 @@
+    ~HostlistConstraint () { hostlist_destroy (hl); };
+
+private:
+    struct hostlist * hl = nullptr;
```
Missed this whitespace in the rebase.

milroy reviewed (changes requested):
Thanks for the quick fixes. Just one space missed and this is good to go. I also asked a question about a shared pointer to improve my understanding of the design, but we can discuss that at a coffee time.

grondo commented on resource/libjobspec/constraint.cpp:
```diff
@@ -0,0 +1,217 @@
+        throw parse_error (operands, msg.c_str ());
+    }
+
+    if (operation == "properties")
+        result = new PropertyConstraint (operands);
+    else if (operation == "hostlist")
+        result = new HostlistConstraint (operands);
+    else if (operation == "ranks")
+        result = new RankConstraint (operands);
+    else if (operation == "and" || operation == "or" || operation == "not")
+        result = new ConditionalConstraint (operation, operands);
+    else {
+        std::string msg = "unknown constraint operator: " + operation;
+        throw parse_error (constraint, msg.c_str ());
+    }
+    return std::shared_ptr<Constraint> (result);
```
Ok, here's the problem: The constraint factory function `constraint_parser()` needs to return a `shared_ptr` because Fluxion needs to store the Constraint in both the Jobspec object for a job and the `jobmeta_t` which is actually used in the traverser. From brief research online, it doesn't make sense to convert a `shared_ptr` to a `unique_ptr` (is there even a way to do it?), so I had to make any vector of Constraints a vector of shared_ptr.

Keep in mind I basically learned all this stuff last week, so I could be missing something obvious. For now I'll use `make_shared` as suggested above (I didn't know about that function, sorry!).

grondo replied:
@trws, your code above worked great. Thanks. I've rebased and squashed and forced a push.

trws replied:
Ok, if it needs to be stored in two places at any time then it needs to be `shared_ptr` there for sure.  Something to know though is that `unique_ptr` was designed to be a good return value for factory functions.  It’s easy, and reasonably cheap, to convert from a `unique_ptr` to a `shared_ptr` so that a factory can always return `unique_ptr` and let the caller decide later.  The conversion should literally be:

```c++
std::shared_ptr<Constraint> c = std::move(constraint_factory(…));
```

Or otherwise get the `unique_ptr` and move it into a same-typed `shared_ptr`.

grondo replied:
Thanks! I will try that. Only the outermost Constraint object needs to be shared I believe.

Hmm, I switched `constraint_parser()` to return `std::unique_ptr<Constraint>`, but existing callers that assigned the result to a shared_ptr still compile without error.

Any reason why this shouldn't throw a compiler error?

```c++
std::unique_ptr<Constraint> constraint_parser (const YAML::Node &constraint);
```

caller:

```c++
std::shared_ptr<Constraint> c = constraint_parser (YAML::Load (t->json));
```

grondo replied:
Ah: https://www.nextptr.com/question/qa1257940863/converting-unique_ptr-to-shared_ptr-factory-function-example

That works out well!

grondo replied:
Ok, @trws and @milroy. I updated the return value from the `constraint_parser()` factory function to a `unique_ptr` along with the type of the values vector in the ConditionalConstraint class. Now the only shared_ptr is on the outermost Constraint object stored in the Jobspec object shared with `jobmeta_t`.

Thanks for the suggestion @trws!

trws replied:
Excellent! Thanks for pushing through this @grondo, should be good both for performance and correctness.  Sorry for the misdirection on the move around the call by the way, I wasn’t thinking about the fact the call already returns an rvalue.

grondo commented:
I believe I have fixed up the remaining whitespace errors, and have updated `constraint_parser()` implementation to use `make_shared` using the suggested code from @trws. Also, now that flux-framework/flux-core#4897 is merged, I added a modicum of end-to-end testing of the more complex RFC 31 job constraints to the testsuite.

trws commented on resource/libjobspec/test/constraint.cpp:
```diff
@@ -0,0 +1,291 @@
+struct match_test match_tests[] = {
+    { "empty json object matches everything", "{}", true },
+    {  "empty properties dict matches everything",
+       "{\"properties\": [] }",
```
It isn’t worth re-doing this for this PR, but for future reference C++11 has raw strings, such that `R”<pat>()<pat>”` where `<pat>` can be any set of zero or more characters such that `)<pat>` does not appear in the body of the string. It makes escapes unnecessary in the string, including across newlines somewhat like `”””` in python or `[[` in lua, so this for example could be `R”({"properties": [] })”`. Not super bad here, but embedded json can get gnarly fast without it.  This is another example from below:

```c++
R"({"and": [ {"properties": ["xx"]},
                      {"properties": ["yy"]}
                    ]})",
```

grondo replied:
Wow, that's really awesome.

grondo commented:
Thanks @trws! I'll wait for @milroy to approve then set MWP.

milroy commented:
LGTM @grondo! Setting MWP.

grondo commented:
Thanks @milroy and @trws!
